src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int INITIAL_CAPACITY = 10;
+    private int capacity;
+    private int size;
+    private Object[] elements;
+
+    ArrayList() {
+        elements = new Object[INITIAL_CAPACITY];
+        capacity = INITIAL_CAPACITY;
+        size = 0;
+    }
+
+    private void resize() {
+        Object[] temp = new Object[size + (size >> 1)];
+        System.arraycopy(elements, 0, temp, 0, elements.length);
+        elements = temp;
+        capacity = size + (size >> 1);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size() == capacity) {
+            resize();
+        }
+        if (isEmpty()) {
+            elements[0] = value;
+            size++;
+        } else {
+            elements[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bounds of the list size");
+        }
+        if (size == capacity) {
+            resize();
+        }
+        size++;
+        for (int i = size - 1; i > index; i--) {
+            elements[i] = elements[i - 1];
+        }
+        elements[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size() - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bounds of the list size");
+        }
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size() - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bounds of the list size");
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is out of bounds of the list size");
+        }
+        T removedelement = (T) elements[index];
+        if (index == size - 1) {
+            size--;
+            return removedelement;
+        }
+        System.arraycopy(elements, index + 1, elements, index,size - index);
+        size--;
+        return removedelement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        int counter = 0;
+        for (Object o : elements) {
+            if (o != null && o.equals(element) || o == element) {
+                index = counter;
+                break;
+            }
+            counter++;
+        }
+        if (index == - 1) {
+            throw new NoSuchElementException();
+        }
+        return remove(index);
```

Comment:
I think this logic is too complex, you can use a for loop and return remove(index) if the element is present

Reply:
it's the only method that has such Exception. remove(index) doesn't.

Reply:
that's why separate if statement for remove(element)